go/chat/boxer.go

```diff
@@ -484,34 +518,49 @@ func (b *Boxer) verify(data []byte, si chat1.SignatureInfo, prefix libkb.Signatu
 	return (err == nil)
 }
 
-// validSenderKey checks that the key is active for sender at ctime.
-func (b *Boxer) validSenderKey(ctx context.Context, sender gregor1.UID, key []byte, ctime gregor1.Time) (bool, libkb.ChatUnboxingError) {
+// ValidSenderKey checks that the key was active for sender at ctime.
+// This trusts the server for ctime, so a colluding server could use a revoked key and this check wouldn't notice.
+// Returns (validAtCtime, revoked, err)
+func (b *Boxer) ValidSenderKey(ctx context.Context, sender gregor1.UID, key []byte, ctime gregor1.Time) (bool, bool, libkb.ChatUnboxingError) {
 	kbSender, err := keybase1.UIDFromString(hex.EncodeToString(sender.Bytes()))
 	if err != nil {
-		return false, libkb.NewPermanentChatUnboxingError(err)
+		return false, false, libkb.NewPermanentChatUnboxingError(err)
 	}
 	kid := keybase1.KIDFromSlice(key)
-	t := gregor1.FromTime(ctime)
+	ctime2 := gregor1.FromTime(ctime)
 
-	var uimap *utils.UserInfoMapper
-	ctx, uimap = utils.GetUserInfoMapper(ctx, b.kbCtx)
-	user, err := uimap.User(kbSender)
-	if err != nil {
-		return false, libkb.NewTransientChatUnboxingError(err)
+	cachedUserLoader := b.kbCtx.GetCachedUserLoader()
+	if cachedUserLoader == nil {
+		return false, false, libkb.NewTransientChatUnboxingError(fmt.Errorf("no CachedUserLoader available in context"))
 	}
-	ckf := user.GetComputedKeyFamily()
-	if ckf == nil {
-		return false, libkb.NewPermanentChatUnboxingError(errors.New("no computed key family"))
-	}
-	activeKey, _, err := ckf.FindActiveSibkeyAtTime(kid, t)
+
+	found, revokedAt, err := cachedUserLoader.CheckKIDForUID(kbSender, kid)
 	if err != nil {
-		return false, libkb.NewPermanentChatUnboxingError(err)
+		return false, false, libkb.NewTransientChatUnboxingError(err)
 	}
-	if activeKey == nil {
-		return false, nil
+	if !found {
+		return false, false, nil
 	}
 
-	return true, nil
+	var revokedAt2 *time.Time
+	if revokedAt != nil {
+		if revokedAt.Unix.IsZero() {
+			return false, false, libkb.NewPermanentChatUnboxingError(fmt.Errorf("zero clock time on expired key"))
+		}
+		t := b.keybase1KeybaseTimeToTime(*revokedAt)
+		revokedAt2 = &t
+	}
+
+	revoked := revokedAt2 != nil
+	validAtCtime := found && (!revoked || revokedAt2.After(ctime2))
```

Comment:
I'm still confused why `found` is used here at all. Isn't it always true if we make it here?